File: modules/expo-vault/android/src/main/java/expo/modules/vault/PdfPages.kt
package expo.modules.vault

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import java.io.File
import java.io.FileOutputStream

/**
 * Rasterises PDF pages to JPEG files so the viewer only ever deals with
 * images. Rendering happens from a decrypted temporary copy and the output
 * lives in the cache directory next to the other decrypted previews.
 */
object PdfPages {
    fun render(source: File, maxPixelSize: Int, directory: File): List<Map<String, Any>> {
        directory.mkdirs()
        val pages = mutableListOf<Map<String, Any>>()
        ParcelFileDescriptor.open(source, ParcelFileDescriptor.MODE_READ_ONLY).use { descriptor ->
            PdfRenderer(descriptor).use { renderer ->
                for (index in 0 until renderer.pageCount) {
                    renderer.openPage(index).use { page ->
                        val scale = maxPixelSize.toFloat() / maxOf(page.width, page.height, 1)
                        val width = maxOf(1, (page.width * scale).toInt())
                        val height = maxOf(1, (page.height * scale).toInt())
                        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
                        bitmap.eraseColor(Color.WHITE)
                        page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
                        val file = File(directory, "page_${index + 1}.jpg")
                        FileOutputStream(file).use { output ->
                            bitmap.compress(Bitmap.CompressFormat.JPEG, 85, output)
                        }
                        bitmap.recycle()
                        pages.add(mapOf("uri" to "file://${file.absolutePath}", "width" to width, "height" to height))
                    }
                }
            }
        }
        return pages
    }
}

File: modules/expo-vault/ios/PdfPages.swift
import Foundation
import PDFKit
import UIKit

enum PdfPagesError: LocalizedError {
  case unreadable

  var errorDescription: String? {
    switch self {
    case .unreadable: return "Could not open the PDF"
    }
  }
}

/// Rasterises PDF pages to JPEG files so the viewer only ever deals with
/// images. Rendering happens from a decrypted temporary copy and the output
/// lives in the cache directory next to the other decrypted previews.
enum PdfPages {
  static func render(from source: URL, maxPixelSize: Int, into directory: URL) throws -> [[String: Any]] {
    guard let document = PDFDocument(url: source) else { throw PdfPagesError.unreadable }
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    var pages: [[String: Any]] = []
    for index in 0..<document.pageCount {
      guard let page = document.page(at: index) else { continue }
      let bounds = page.bounds(for: .mediaBox)
      let scale = CGFloat(maxPixelSize) / max(bounds.width, bounds.height, 1)
      let size = CGSize(width: max(1, bounds.width * scale), height: max(1, bounds.height * scale))
      let format = UIGraphicsImageRendererFormat.default()
      format.scale = 1
      let image = UIGraphicsImageRenderer(size: size, format: format).image { context in
        UIColor.white.setFill()
        context.fill(CGRect(origin: .zero, size: size))
        context.cgContext.translateBy(x: 0, y: size.height)
        context.cgContext.scaleBy(x: scale, y: -scale)
        context.cgContext.translateBy(x: -bounds.origin.x, y: -bounds.origin.y)
        page.draw(with: .mediaBox, to: context.cgContext)
      }
      guard let jpeg = image.jpegData(compressionQuality: 0.85) else { continue }
      let url = directory.appendingPathComponent("page_\(index + 1).jpg")
      try jpeg.write(to: url, options: .atomic)
      pages.append(["uri": url.absoluteString, "width": Int(size.width), "height": Int(size.height)])
    }
    return pages
  }
}
